### ISSUE TITLE
Try to link with -lfoo when find_library(libfoo) was used

### ISSUE THREAD

**Issue by martinetd** [mesonbuild/meson#3537]
Hi,

We ran into an issue at swaywm/sway#1920 where we used `cc.find_library(libpam)`, which works on most systems but will fail on some systems e.g. nixos.

The code for `find_library` first tries to link with `-l<libname>` then does `get_library_dirs` (for gcc, `--print-search-dirs` libraries' line) and manually looks for `lib<libname>` and `<libname>` with various suffixes there.

Would if make sense to add something like the following, to try to link with `-l<libname-without-starting-lib>` if such an argument was given? (I can submit a PR if someone here agrees)
```diff
diff --git a/mesonbuild/compilers/c.py b/mesonbuild/compilers/c.py
index 3f9ba5cc..28a7a61c 100644
--- a/mesonbuild/compilers/c.py
+++ b/mesonbuild/compilers/c.py
@@ -724,6 +724,10 @@ class CCompiler(Compiler):
             args = ['-l' + libname]
             if self.links(code, env, extra_args=args):
                 return args
+            if libname.startswith('lib'):
+                args = ['-l' + libname[3:]]
+                if self.links(code, env, extra_args=args):
+                    return args
         # Not found? Try to find the library file itself.
         extra_dirs += self.get_library_dirs()
         suffixes = ['so', 'dylib', 'lib', 'dll', 'a']
```

As a small note, nixos uses a gcc wrapper to actually find the lib and looks for CFLAGS/LDFLAGS in a non-standard place, but manually setting LDFLAGS to `-L<dirwherelibpamis>` does not lead to that path being added to `get_library_dirs()` so fails anyway; I don't think it'd be wise to scrap CFLAGS/LDFLAGS for directories to consider using manually but that means there is no workaround for this (aside of, well, fixing the meson.build!)

**Comment by nirbheek** [comment]
`cc.find_library('libfoo')` works because it used to work by accident. We should warn when people pass `libfoo` to it. Would that be enough for NixOS? Because then at least upstreams won't ask you to explain. :)

**Comment by martinetd** [comment]
I'm not fussy either way.
I agree it works by accident more than anything else so happy with a warning instead - what should the exact criteria be, though? Found through the manual loop + startswith 'lib'?

Just looking at name isn't enough (because of libs like liblibreoffice that start with lib..); and I guess the search loop has its uses so we probably don't want to systematically warn when we fall back to it...

**Comment by jpakkane** [comment]
You should use `find_library('pam')`. Nothing else is portable, because on Windows libraries do not have a `lib` prefix. This is consistent with all other usages, such as looking up the math library with `find_library('m')` rather than `find_library('libm')`.

**Comment by martinetd** [comment]
Yes, we have already updated sway to do that at the same time as opening this issue.
I'd still like to see some form of warning as @nirbheek suggested, happy to do it if the criteria I suggested seems good enough.

**Comment by nirbheek** [comment]
I searched for libraries that start with `liblib` and what I saw next will shock you!

```console
# dnf repoquery --whatprovides '/usr/lib64/liblib*'
engrid-0:2.0.0-0.27.20170615git0563bcc.fc28.x86_64
engrid-devel-0:2.0.0-0.27.20170615git0563bcc.fc28.x86_64
kremotecontrol-libs-0:17.08.3-3.fc28.x86_64
libreofficekit-1:6.0.3.2-10.fc28.x86_64
```
`kremotecontrol-libs` is a runtime-only library with no -devel. However, libreofficekit and engrid do:

```
/usr/lib64/liblibreofficekitgtk.so
/usr/lib64/liblibengrid.so
```

For libengrid, it looks like a packaging mistake, and we should rightfully warn about it.

For libreoffice, it would be a false positive, so we should explicitly exclude that from the warning.

**Comment by martinetd** [comment]
Yeah, I did a similar search back then, but I'd rather not have an explicit exclusion list like that (as far as my personal taste is involved at least) - hence my idea of only warning if -lfoo didn't work and foo starts with lib.

For example, liblibreofficekitgtk is in its proper place so it would supposedly be found just fine with -llibreofficekitgtk, and there would be no reason to print that warning in this case... But trying to search for libpam would fail -llibpam and find the file so that could cause a warning as libpam starts with lib.
I'm not sure how generic enough it would be for other OSes though.


(I'm actually surprised there aren't more libraries starting with lib, there's plenty of words to pick from there. . . like libanotophorous or something :P)

**Comment by QuLogic** [comment]
Might want to check [Debian](https://packages.debian.org/search?suite=experimental&arch=any&mode=filename&searchon=contents&keywords=liblib) too?

**Comment by bonzini** [comment]
Closing, and opening a separate issue for the `liblibreofficegtk.so` false positive.
